add native tunnel encapsulation support for the XDP Loadbalancer

julianwiedmann opened the pull request:
https://github.com/cilium/cilium/pull/21364 added support for combining XDP with tunnel-routing. This was done through a fallback mechanism, where we would (1) punt packets from XDP up to the `from-netdev` TC program, and (2) redirect them to `cilium_vxlan` / `cilium_geneve` as usual.

Besides the performance implications of the additional hop through the TC path, this also meant that https://github.com/cilium/cilium/pull/23890 later couldn't add support for IPv6 traffic with DSR-GENEVE using XDP (as there was no easy way of transporting the IPv6 DSR information from XDP all the way to the TC encap path).

This PR introduces native XDP support for adding VXLAN / GENEVE tunnel headers to a packet, and thus avoids the detour through the TC path.

open aspects
- ~~select a proper per-flow OuterSrcPort. It's not a hard blocker to get all the core infrastructure merged, so I plan to tackle this in a follow-on PR.~~ https://github.com/cilium/cilium/pull/26177
- ~~the `send_trace_notify()` in `__encap_with_nodeid()` happens too late for the XDP path - we want it to describe the inner packet, but it already has the tunnel headers. Easy enough to sort out in a follow-on PR.~~ https://github.com/cilium/cilium/pull/25828

julianwiedmann committed:
bpf: encap: manually set src IP and port from nodeport XDP paths

In preparation for manually building the tunnel headers from XDP context,
pass the outer header's Src IP and Port to the encap helpers.

Set the OuterSrcIP as IPV4_DIRECT_ROUTING. For the OuterSrcPort we're
simply re-using TUNNEL_PORT for now, but this should eventually turn into
a hash of the flow's 5-tuple
(see https://www.rfc-editor.org/rfc/rfc8926.html#name-udp-header).

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
datapath: set correct TUNNEL_PROTOCOL for DSR-GENEVE ad-hoc tunnel

For DSR-GENEVE with native routing we need the same special handling as
when Reinitialize() prepares the parameters for the init.sh script.
Otherwise we most likely set it to TUNNEL_VXLAN (the default), when we
actually want TUNNEL_GENEVE.

We can look into consolidating these duplicated checks at a later time.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: xdp: add native encap support

When the XDP nodeport code requires tunnel encapsulation, we currently push
the packet up to the from-netdev TC program (with XFER_PKT_ENCAP set) and
let it handle the tunnel redirect.

Replace this with an in-XDP implementation that adds the outer headers
for VXLAN or GENEVE. This path now also returns CTX_ACT_REDIRECT, but in
contrast to the skb implementation it doesn't set the `ifindex`. We handle
this case in the relevant nodeport paths, and fall back to an IPv4 FIB
lookup for the outer packet.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: xdp: bump supported length for ctx_adjust_hroom()

When adding GENEVE encapsulation with IPv6 DSR Option, we need some
additional space.

Note that most drivers should offer XDP_PACKET_HEADROOM (256B), but some
only provide 192B.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: remove XDP-to-TC punt infrastructure for tunnel

Now that we have XDP support for adding tunnel headers, nothing sets
XFER_PKT_ENCAP anymore. Remove all the relevant code and parameters.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
test: service: enable IPv6 in XDP + dsr-geneve test

With the in-XDP encap support, the IPv6 path should now also work.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
datapath: require direct routing device for XDP with tunnel routing

When using XDP-native encap, we won't go via the cilium_* tunnel device.
To build the tunnel headers, we need IPV4_DIRECT_ROUTING to be populated.

So require the direct routing device, even when tunnel-routing is enabled.
Also add a test for this scenario.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: nodeport: fix up indentation for GENEVE-DSR encap calls

Make it a bit easier to work with this in subsequent patches.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
docs: remove DSR-GENEVE restriction for IPv6 in XDP

With in-XDP encap support, this no longer applies.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

zacharysarah reviewed (approved):
LGTM from a docs perspective.

asauber reviewed (approved):
Config-related changes LGTM

ysksuzuki commented on bpf/lib/overloadable_xdp.h:
```diff
@@ -184,35 +159,88 @@ static __always_inline bool ctx_snat_done(struct xdp_md *ctx)
+			if (opt_len > 0)
+				memcpy((void *)geneve + sizeof(*geneve), opt, opt_len);
+
+			geneve->opt_len = (__u8)(opt_len >> 2);
```
I think we can set the critical bit in the Geneve header, right?

Here is the pcap of a Geneve packet (I see that the source port is fixed, but I assume that this will be addressed in a follow-on.)
<img width="657" alt="geneve-xdp" src="https://github.com/cilium/cilium/assets/5572688/971fb5fc-d0bd-4210-abb1-bf53f7f8a2d1">

julianwiedmann replied:
If I remember correctly, the conclusion from https://github.com/cilium/cilium/pull/23890 was that we currently can't set the critical bit from the TC path - correct? Then let's discuss this aspect in a follow-on PR - imho it's not clear whether we want different behaviour from XDP and TC tunneling.

ysksuzuki replied:
Correct. There's no kernel API to set the critical bit from the TC path for now. I thought we could set it by ourselves here, but year, let's discuss it in a follow-on.

julianwiedmann replied:
I've parked a WIP [here](https://github.com/cilium/cilium/pull/25959), so we don't lose track of it

ysksuzuki commented on bpf/lib/encap.h:
```diff
@@ -233,11 +235,32 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
 }
 #endif /* TUNNEL_MODE */
 
+static __always_inline __be16 tunnel_gen_src_port_v4(void)
```
Are you planning to implement this as a follow-on?

Does this function generate a source port using the flow hash like that?
https://github.com/torvalds/linux/blob/0d85b27b0cc6b5cf54567c5ad913a247a71583ce/include/net/udp.h#L206

julianwiedmann replied:
Yep, just a detail that can follow afterwards once all the core infrastructure is sorted.

borkmann replied:
This is for RSS steering on remote nodes, right? Perhaps rss_gen_src{4,6}() could be reused given we also hash in info from inner pkt.

julianwiedmann replied:
> This is for RSS steering on remote nodes, right? 

yeah, the Geneve RFC says for instance:

> This source port SHOULD be the same for all packets belonging to a single encapsulated flow to prevent reordering due to the use of different paths. To encourage an even distribution of flows across multiple links, the source port SHOULD be calculated using a hash of the encapsulated packet headers using, for example, a traditional 5-tuple.

So I'll definitely have that follow-on PR in before `v1.14` ships.
 

>Perhaps rss_gen_src{4,6}() could be reused given we also hash in info from inner pkt.

Pretty much, yes (except that it needs to spit out a `__be16`). We also have `hash_from_tuple_v4()` :).

julianwiedmann replied:
Playing around in https://github.com/cilium/cilium/pull/26177.

ysksuzuki reviewed (changes requested):
Thank you for the great work!  Just a few comments from a quick look and tested it in my environment.  It works great!

julianwiedmann committed:
bpf: set TUNNEL_PROTOCOL_GENEVE for XDP complexity tests with DSR-GENEVE

Configs with DSR-GENEVE will always use a GENEVE tunnel. So make it
possible to override the default (VXLAN), and use the correct macro for
the complexity tests.

This codepath is slightly more complex (it needs to handle GENEVE options),
so ensure that we cover it in the tests.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: extend compile / complexity tests for native XDP tunnel encap

Cover both VXLAN and GENEVE in the XDP compile tests. And add some
additional complexity configs for XDP that cover tunnel-routing.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

jibi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -140,7 +140,15 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	}
 
-	cDefinesMap["TUNNEL_PROTOCOL"] = fmt.Sprintf("%d", tunnelProtocols[option.Config.TunnelProtocol])
+	encapProto := option.Config.TunnelProtocol
```
should this live in daemon_main.go (and perhaps log an info with the actual tunneling protocol used)?

julianwiedmann replied:
yeah, I'd love to clean this up (see the commit message for this code). We already have multiple copies of the "special handling for ad-hoc tunnel": https://github.com/cilium/cilium/blob/6b84bef94ba46ed3c09bf7a242a711449f3003bd/pkg/datapath/loader/base.go#L397 https://github.com/cilium/cilium/blob/6b84bef94ba46ed3c09bf7a242a711449f3003bd/pkg/option/config.go#L3141

But over-riding the `config.TunnelProtocol` at a high level means teaching all existing users to tolerate `TunnelDisabled`. And I'd rather do that in a separate PR ... :)

rgo3 replied:
I've opened an issue for a follow-up, hope there wasn't one already.

julianwiedmann replied:
This is https://github.com/cilium/cilium/issues/25769. Thanks Robin!

qmonnet commented on test/k8s/services.go:
```diff
@@ -561,7 +561,7 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`,
 			})
 			// DSR with Geneve doesn't work in conjunction with XDP and IPv6 currently
-			testNodePortExternalIPv4Only(kubectl, ni, false, true, true)
+			testNodePortExternal(kubectl, ni, false, true, true)
```
Can you please update the comment here?

qmonnet reviewed (approved):
Looks all good to me, with just one nit on a comment that would need an update (see below). Nice work!

rgo3 reviewed (approved):
LGTM for loader related changes.
